impacket/dcerpc/v5/tsts.py

```diff
@@ -0,0 +1,3721 @@
+'''
+BOOLEAN RpcRemoteAssistancePrepareSystemRestore(
+ [in] SERVER_HANDLE hServer,
+ [out] DWORD* pResult
+);
+'''
+# 3.7.4.1.34 RpcWinStationGetAllProcesses_NT6 (Opnum 70)
+'''
+BOOLEAN RpcWinStationGetAllProcesses_NT6(
+ [in] SERVER_HANDLE hServer,
+ [out] DWORD* pResult,
+ [in] ULONG Level,
+ [in, out] BOUNDED_ULONG* pNumberOfProcesses,
+ [out, size_is(,*pNumberOfProcesses)]
+ PTS_ALL_PROCESSES_INFO_NT6* ppTsAllProcessesInfo
+);
+'''
+
+#OLD 3.6.4.1.54 RpcWinStationRegisterNotificationEvent (Opnum 71)
+'''
+BOOLEAN RpcWinStationRegisterNotificationEvent(
+ [in] SERVER_HANDLE hServer,
+ [out] DWORD* pResult,
+ [out] REGISTRATION_HANDLE* pNotificationId,
+ [in] ULONG_PTR EventHandle,
+ [in] DWORD TargetSessionId,
+ [in] DWORD dwMask,
+ [in] DWORD dwProcessId
+);
+'''
+#OLD 3.6.4.1.55 RpcWinStationUnRegisterNotificationEvent (Opnum 72)
+'''
+BOOLEAN RpcWinStationUnRegisterNotificationEvent(
+ [in] SERVER_HANDLE hServer,
+ [out] DWORD* pResult,
+[in, out] REGISTRATION_HANDLE* NotificationId
+);
+'''
+
+#OLD 3.6.4.1.56 RpcWinStationAutoReconnect (Opnum 73)
+'''
+BOOLEAN RpcWinStationAutoReconnect(
+ [in] SERVER_HANDLE hServer,
+ [out] DWORD* pResult,
+ [in] DWORD LogonId,
+ [in] DWORD flags
+);
+'''
+#OLD 3.6.4.1.57 RpcWinStationCheckAccess (Opnum 74)
+'''
+BOOLEAN RpcWinStationCheckAccess(
+ [in] SERVER_HANDLE hServer,
+ [out] DWORD* pResult,
+ [in] DWORD ClientLogonId,
+ [in] DWORD UserToken,
+ [in] ULONG LogonId,
+ [in] ULONG AccessMask
+);
+'''
+
+# 3.7.4.1.35 RpcWinStationOpenSessionDirectory (Opnum 75)
+# Does not work
+class RpcWinStationOpenSessionDirectory(NDRCALL):
+    opnum = 75
+    structure = (
+        ('hServer', SERVER_HANDLE),
+        ('pszServerName', WSTR),
+    )
+
+class RpcWinStationOpenSessionDirectoryResponse(NDRCALL):
+    structure = (
+        ('pResult', pResult_ENUM),
+        ('ErrorCode', BOOLEAN),
+    )
+
+
+################################################################################
+# Helper Functions
+################################################################################
+
+class TSTSEndpoint:
+    def __init__(self, smb, target_ip, stringbinding, endpoint, kerberos = False):
+        self._stringbinding = stringbinding.format(target_ip)
+        self._endpoint = endpoint
+        self._smbconnection = smb
+        self._bind()
+    def _bind(self):
+        self._rpctransport = transport.DCERPCTransportFactory(self._stringbinding)
+        self._rpctransport.set_smb_connection(self._smbconnection)
+        self._dce = self._rpctransport.get_dce_rpc()
+        self._dce.set_credentials(*self._rpctransport.get_credentials())
+        self._dce.connect()
+        self._dce.set_auth_level(RPC_C_AUTHN_LEVEL_PKT_PRIVACY)
+        self._dce.bind(self._endpoint)
+        return self._dce
+    def _disconnect(self):
+        self._dce.disconnect()
+    def __enter__(self):
+        return self
+    def __exit__(self, type, value, traceback):
+        self._disconnect()
```

Comment:
While I like the approach of having the class initializing the connection, it's not the pattern we've been following in the rest of the DCE/RPC endpoints. Currently, the helper functions are just abstractions to avoid filling the structures' fields individually.

For this PR and for consistency-sake, I'd suggest we update the helpers to just follow the current approach and we can discuss/review if does make sense to have class-level helpers moving forward in a separate PR.

Reply:
I would like to keep `TSTSEndpoint` initialization class along with keeping consistency of helpers functions.
This approach significantly decrease example's code, and i hope it will be accepted.